### PR TITLE
Added: Implementation of neighbourhood blocking

### PR DESCRIPTION
**Pull request by dan-elias**
## Purpose

This branch adds an implementation of Neighbourhood Blocking - an indexing algorithm which generalizes Sorted Neighbourhood Indexing by allowing:
* matching by *simultaneous* proximity in multiple record orderings
* wildcard matching of null values
* a limited number of complete mismatches (eg: match any 3 fields out of 5)

The algorithm is described in [this paper](https://github.com/dan-elias/neighbourhood_blocking/blob/master/ADC19_full_08.pdf) which will be presented at the [Australasian Database Conference](http://www.acsw.org.au/) next month.

## Summary of changes

* New class - NeighbourhoodBlock (subclass of Block) - added to recordlinkage.index
* New method - neighbourhoodblock added to recordlinkage.api.Index
* Unit tests in tests/test_indexing.py:
  - NeighbourhoodBlock instance added to output of get_test_algorithms
  - class TestNeighbourhoodBlock added
* Mentions of NeighbourhoodBlock added to documentation in docs/ref-index.rst

### CODE REVIEW

**Comment by dan-elias** [comment]
Tests failed on some versions.  Will correct and resubmit.

**Comment by dan-elias** [comment]
Tests passing for all configurations except the two outlined in the table below

|                                Job log                                | Python vers | Pandas vers |
|-----------------------------------------------------------------------|-------------|-------------|
| [394.3](https://travis-ci.org/J535D165/recordlinkage/jobs/473509733)  |         3.6 |        0.18 |
| [394.15](https://travis-ci.org/J535D165/recordlinkage/jobs/473509745) |         3.6 |        0.22 |

In both cases, the following error is raised when pandas is imported:
```
pandas/src/numpy.pxd:865: in init pandas.hashtable (pandas/hashtable.c:38368)
    ???
E   ValueError: numpy.ufunc has the wrong size, try recompiling. Expected 192, got 216
```

As far as I can see, this seems to be a version incompatibility between pandas and numpy in those particular configs (tests passed in other configs using python 3.6 or either of those versions of pandas).

File: recordlinkage/api.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 from recordlinkage.base import BaseIndex, BaseCompare
-from recordlinkage.index import Full, Block, SortedNeighbourhood, Random
+from recordlinkage.index import Full, Block, SortedNeighbourhood, NeighbourhoodBlock, Random
```
Comment (J535D165) [inline recordlinkage/api.py]: Remove changes in this file.

File: recordlinkage/index.py [hunk 2]
```diff
@@ -459,6 +459,175 @@ def _dedup_index(self, df_a):
             levels=levels, labels=labels, verify_integrity=False)
 
 
+class NeighbourhoodBlock(Block):
```
Comment (J535D165) [inline recordlinkage/index.py]: Move this class to a new Python file (for example neighbourhoodblock.py) in the contrib folder https://github.com/J535D165/recordlinkage/tree/master/recordlinkage/contrib/index.

File: tests/test_indexing.py [hunk 3]
```diff
@@ -647,3 +650,123 @@ def test_random_with_replace(self):
         pairs2 = index_cl2.index(self.a)
         assert len(pairs2) == 1000
         assert not pairs2.is_unique
+
+
+# NeighbourhoodBlock
+
+class TestNeighbouroodBlock(TestData):
```
Comment (J535D165) [inline tests/test_indexing.py]: Move this class to a new Python file (neighbourhoodblock_test.py) in the contrib folder. 

Check out https://github.com/J535D165/recordlinkage/tree/master/recordlinkage/contrib/compare for inspiration.

Comment (J535D165) [inline tests/test_indexing.py]: Typo in class name: TestNeighbour**h**oodBlock

File: tests/test_indexing.py [hunk 4]
```diff
@@ -647,3 +650,123 @@ def test_random_with_replace(self):
...
+                vals = vals.copy()
+                vals.iloc[np.random.choice(len(df), size=nan_count, replace=False)] = np.nan
+                return vals
+            return df.copy() if nan_count <= 0 else df.apply(with_nulls)
+
+        np.random.seed(0)
+        cls.incomplete_a = incomplete_df_copy(cls.a)
+        cls.incomplete_b = incomplete_df_copy(cls.b)
+
+    def assert_index_comparisons(self, pairwise_comparison, indexers, *args, **kwargs):
```
Comment (J535D165) [inline tests/test_indexing.py]: Really nice. I think this can be useful for other tests as well. Thanks!

File: docs/ref-index.rst [hunk 5]
```diff
@@ -23,6 +23,7 @@ references for background information about indexation.
     .. automethod:: recordlinkage.Index.full
     .. automethod:: recordlinkage.Index.block
     .. automethod:: recordlinkage.Index.sortedneighbourhood
+    .. automethod:: recordlinkage.Index.neighbourhoodblock
```
Comment (J535D165) [inline docs/ref-index.rst]: I'm not completely sure on what to do with the documentation on the contrib module. I think I will create a new subsection in the docs. For now, removing these changes will help.

**Comment by J535D165** [comment]
Hello Dan-Elias, 

Great paper and great code. I really appreciate your work! I hope you agree with me on the contrib submodule. 

Next week, I'm planning a new release of the package. I hope we can include your algorithm! This gives you a nice opportunity to get your code packaged and shipped before the conference!

Best, Jonathan

**Comment by dan-elias** [comment]
Jonathan,

Thanks for the quick response!

Yes - that's fine. Will you do it, or do you want me to?

- Dan

On 31 Dec. 2018 09:32, Jonathan de Bruin <notifications@github.com> wrote:

@J535D165 requested changes on this pull request.

I'm requesting one change before merging it. I would like to move the code to the new contrib submodule. This was inspired by tensorflow (were it resulted in many cool new products). I hope you are fine with this!

Example code:

from recordlinkage.contrib.index import NeighbourhoodBlock

indexer = NeighbourhoodBlock(keys, windows=windows, max_nulls=1, max_non_matches=2)
indexer.index(census_data_1980, census_data_1990)

or

from recordlinkage.contrib.index import NeighbourhoodBlock

indexer = recordlinkage.Index()
inderer.add(NeighbourhoodBlock(keys, windows=windows, max_nulls=1, max_non_matches=2))
indexer.index(census_data_1980, census_data_1990)

________________________________

In recordlinkage/api.py<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546410>:

> @@ -1,5 +1,5 @@
 from recordlinkage.base import BaseIndex, BaseCompare
-from recordlinkage.index import Full, Block, SortedNeighbourhood, Random
+from recordlinkage.index import Full, Block, SortedNeighbourhood, NeighbourhoodBlock, Random


Remove changes in this file.

________________________________

In recordlinkage/index.py<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546517>:

> @@ -459,6 +459,175 @@ def _dedup_index(self, df_a):
             levels=levels, labels=labels, verify_integrity=False)


+class NeighbourhoodBlock(Block):


Move this class to a new Python file (for example neighbourhoodblock.py) in the contrib folder https://github.com/J535D165/recordlinkage/tree/master/recordlinkage/contrib/index.

________________________________

In tests/test_indexing.py<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546554>:

> @@ -647,3 +650,123 @@ def test_random_with_replace(self):
         pairs2 = index_cl2.index(self.a)
         assert len(pairs2) == 1000
         assert not pairs2.is_unique
+
+
+# NeighbourhoodBlock
+
+class TestNeighbouroodBlock(TestData):


Move this class to a new Python file (neighbourhoodblock_test.py) in the contrib folder.

Check out https://github.com/J535D165/recordlinkage/tree/master/recordlinkage/contrib/compare for inspiration.

________________________________

In tests/test_indexing.py<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546566>:

> @@ -647,3 +650,123 @@ def test_random_with_replace(self):
         pairs2 = index_cl2.index(self.a)
         assert len(pairs2) == 1000
         assert not pairs2.is_unique
+
+
+# NeighbourhoodBlock
+
+class TestNeighbouroodBlock(TestData):


Typo in class name: TestNeighbourhoodBlock

________________________________

In tests/test_indexing.py<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546594>:

> +        TestData.setup_class()
+
+        def incomplete_df_copy(df, nan_proportion=0.1):
+            'copy of DataFrame with some cells set to NaN'
+            nan_count = int(round(len(df) * nan_proportion))
+            def with_nulls(vals):
+                vals = vals.copy()
+                vals.iloc[np.random.choice(len(df), size=nan_count, replace=False)] = np.nan
+                return vals
+            return df.copy() if nan_count <= 0 else df.apply(with_nulls)
+
+        np.random.seed(0)
+        cls.incomplete_a = incomplete_df_copy(cls.a)
+        cls.incomplete_b = incomplete_df_copy(cls.b)
+
+    def assert_index_comparisons(self, pairwise_comparison, indexers, *args, **kwargs):


Really nice. I think this can be useful for other tests as well. Thanks!

________________________________

In docs/ref-index.rst<https://github.com/J535D165/recordlinkage/pull/81#discussion_r244546642>:

> @@ -23,6 +23,7 @@ references for background information about indexation.
     .. automethod:: recordlinkage.Index.full
     .. automethod:: recordlinkage.Index.block
     .. automethod:: recordlinkage.Index.sortedneighbourhood
+    .. automethod:: recordlinkage.Index.neighbourhoodblock


I'm not completely sure on what to do with the documentation on the contrib module. I think I will create a new subsection in the docs. For now, removing these changes will help.

File: recordlinkage/api.py (hunk 1 above)
Comment (dan-elias) [reply recordlinkage/api.py]: done

File: recordlinkage/index.py (hunk 2 above)
Comment (dan-elias) [reply recordlinkage/index.py]: done

File: tests/test_indexing.py (hunk 4 above)
Comment (dan-elias) [reply tests/test_indexing.py]: Thanks

**Comment by dan-elias** [comment]
Thanks for the quick response!  Please disregard my earlier email.

Changes made as requested with one variation (in unresolved conversation above - location of test_neighbourhoodblock.py).  [Test status](https://travis-ci.org/J535D165/recordlinkage/builds/473865520?utm_source=github_status&utm_medium=notification) is unchanged.

File: tests/test_indexing.py (hunk 3 above)
Comment (dan-elias) [reply tests/test_indexing.py]: Also note:  the NeighbourhoodBlock instance is still included in get_test_algorithms (in tests/test_indexing.py).  Happy to delete it if you wish.

**Pull request merged**